rust/frontend/src/value.rs

```diff
@@ -0,0 +1,540 @@
+        assert_eq!(
+            self.type_code,
+            TypeCode::kDLFloat,
+            "Requires f64, but found {:?}",
+            self.type_code
+        );
+        unsafe { self.value.inner.v_float64 }
+    }
+
+    pub fn to_bytearray(&self) -> TVMByteArray {
+        assert_eq!(
+            self.type_code,
+            TypeCode::kBytes,
+            "Requires byte array, but found {:?}",
+            self.type_code
+        );
+        unsafe {
+            let barr_ptr =
+                mem::transmute::<*mut c_void, *mut ts::TVMByteArray>(self.value.inner.v_handle);
+            TVMByteArray::new(*barr_ptr)
+        }
+    }
+
+    pub fn to_module(&self) -> Module {
+        assert_eq!(
+            self.type_code,
+            TypeCode::kModuleHandle,
+            "Requires module handle, but found {:?}",
+            self.type_code
+        );
+        let module_handle = unsafe { self.value.inner.v_handle };
+        Module::new(module_handle, false, None)
+    }
+
+    pub fn to_string(&self) -> String {
+        assert_eq!(
+            self.type_code,
+            TypeCode::kStr,
+            "Requires string, but found {:?}",
+            self.type_code
+        );
+        let sptr: *const c_char = unsafe { self.value.inner.v_str };
+        let ret_str = unsafe {
+            match CStr::from_ptr(self.value.inner.v_str).to_str() {
+                Ok(s) => s,
+                Err(_) => "Invalid UTF-8 message",
+            }
+        };
+        ret_str.to_string()
+    }
+
+    pub fn to_ndarray(&self) -> NDArray {
+        assert_eq!(
+            self.type_code,
+            TypeCode::kArrayHandle,
+            "Requires Array handle, but found {:?}",
+            self.type_code
+        );
+        let handle = unsafe { self.value.inner.v_handle };
+        let arr_handle = unsafe { mem::transmute::<*mut c_void, ts::TVMArrayHandle>(handle) };
+        NDArray::new(arr_handle, true)
+    }
+
+    pub fn to_type(&self) -> TVMType {
+        assert_eq!(
+            self.type_code,
+            TypeCode::kTVMType,
+            "Requires TVMType, but found {:?}",
+            self.type_code
+        );
+        let ty = unsafe { self.value.inner.v_type };
+        TVMType::from(ty)
+    }
+
+    pub fn to_ctx(&self) -> TVMContext {
+        assert_eq!(
+            self.type_code,
+            TypeCode::kTVMContext,
+            "Requires TVMContext, but found {:?}",
+            self.type_code
+        );
+        let ctx = unsafe { self.value.inner.v_ctx };
+        TVMContext::from(ctx)
+    }
+}
+
+/// Main way to create a TVMArgValue from suported Rust's values.
+impl<'b, 'a: 'b, T: 'b + ?Sized> From<&'b T> for TVMArgValue<'a>
+where
+    TVMValue: From<&'b T>,
+    TypeCode: From<&'b T>,
+{
+    fn from(arg: &'b T) -> Self {
+        TVMArgValue::new(TVMValue::from(arg), TypeCode::from(arg))
+    }
+}
+
+/// TVMRetValue is an alias of TVMArgValue.
+pub type TVMRetValue<'a> = TVMArgValue<'a>;
```

Comment:
TVMRetValue is actually an owned version of TVMArgValue. The runtime implementation captures this, so we should use that. No need to duplicate code.

Reply:
This is something I don't understand. Why do we need an own TVMRetValue? what if I want to input the ret val into another function?

Reply:
https://github.com/dmlc/tvm/blob/master/include/tvm/runtime/packed_func.h#L576

Reply:
I thought about that and that doesn't answer my question from Rust frontend point of view. For example, what happens when TVMRetValue needs to going through another function call? *I'm questioning the above design actually from Rust's point of view*

Specifically in `TVMArgValue`, the rather artificial lifetime, *exist only* to protect the val and type code when going through FFI and this implementation shows that nothing changes if we don't make `TVMRetValue` owned here and the conversion `to_*` methods for `TVMArgValue` would suffice. I can provide `to_owned` method if desired. Furthermore, downcasting from `Any` is the last thing I want to do.

@tqchen Since this is an important concern for me, may I have your opinion on this?

Reply:
I'm trying to incorporate your code, but the problem is I need `Clone` for `TVMRetValue` which is used in `function::Builder` and cannot impl `Clone` for `Any`!

Reply:
> `TVMRetValue` should be `AsRef<TVMArgValue>` then
Could you clarify? you mean as type def `type TVMRetValue = Box<dyn AsRef<TVMArgValue>>` or something?

Reply:
> you mean as ...

No, more like `impl AsRef<TVMArgValue> for TVMRetvalue` if you can swing it.

> impl `Clone` for `TVMRetValue`

You'll need to `impl Clone` and clone the actual data field that's in use for the particular type code. Shouldn't be too difficult.

Reply:
One issue with `Box<Any>` is that in `tests/callback/bin/error`, registering a global function that errors `fn(&[TVMArgValue]) -> Result<TVMRetValue>` results in: 
"the type `(dyn std::any::Any + 'static)` may not be safely transferred across an unwind boundary" in [catch_unwind](https://doc.rust-lang.org/std/panic/fn.catch_unwind.html).

Reply:
Since I'm very hesitant of downcasting from `Box<Any>`, I decided to go the easier approach of having `TVMRetValue` contain the val and type code and providing *explicit* conversion methods same as `TVMArgValue`.

Reply:
It's not really an `Any`, though. `TVMPODValue`--both in Rust and C++ is just a fat pointer. You can safely downcast it once you check that the type code stored in the PODValue matches. This makes the Rust and C++ interfaces consistent.

I agree that compile-time "checking" is helpful, but it's moot when you can't actually guarantee that the fat pointer contains what you think it does (e.g., when returned from the TVM C++ API).

Reply:
well partially agree, but again like to explicit as much as possible, also that would require many other changes which I'm hesitant to go through. Exhausted now :(

Reply:
> Exhausted now

I can imagine. This was a mammoth PR you made. Did you end up implementing `Clone` for `common::TVMRetValue` using the `typeCode`-guided downcast? If not, I can take a whack at that closer to Friday.

Reply:
> This was a mammoth PR you made

:sweat_smile:

> Did you end up implementing Clone for common::TVMRetValue using the typeCode-guided downcast?

No actually, just made 

```
pub struct TVMRetValue {
    value: TVMValue,
    type_code: TypeCode,
```

and the same explicit conversion `to_*`.

> I can take a whack at that closer to Friday

That'd be great :+1: 

> resolve those few issues around cloning vecs

Well, right now only `shape()` creates an owned vec. I changed `stides()` same as you.

Reply:
can you make `shape` return a slice real quick? even if only for consistency, `ndarray` [returns a slice](https://docs.rs/ndarray/0.12.1/ndarray/struct.ArrayBase.html#method.shape)

Reply:
also, in case I end up getting to `TVM*Value` before you, you can please explain what exactly you need of your implementations that isn't satisfied by those in `runtime` so that I can focus my efforts?

Reply:
> what exactly you need of your implementations that isn't satisfied by those in runtime so that I can focus my efforts?

There should be anything I think. Guess the main difficulty for unification is I had to make my own `TVMValue` with `ValueKind` for debugging (as union cannot be debug formatted), so needs proper adjustments and I'm not implementing any automatic `Into` or `TryInto` to be explicit!

Reply:
> can you make shape return a slice real quick?

Done.